pkg/controller/tailoredprofile/tailoredprofile_controller.go

```diff
@@ -536,6 +657,106 @@ func (r *ReconcileTailoredProfile) getRulesFromSelections(tp *cmpv1alpha1.Tailor
 	}
 	return rules, nil
 }
+
+func (r *ReconcileTailoredProfile) getCustomRulesFromSelections(tp *cmpv1alpha1.TailoredProfile) (map[string]*cmpv1alpha1.CustomRule, error) {
+	rules := make(map[string]*cmpv1alpha1.CustomRule, len(tp.Spec.EnableRules)+len(tp.Spec.DisableRules)+len(tp.Spec.ManualRules))
+	ruleKind := cmpv1alpha1.CustomRuleKind
+	for _, selection := range append(tp.Spec.EnableRules, append(tp.Spec.DisableRules, tp.Spec.ManualRules...)...) {
+		// Skip non-CustomRule selections (Kind must be explicitly set to CustomRule)
+		if selection.Kind != ruleKind {
+			continue
+		}
+		_, ok := rules[selection.Name]
+		if ok {
+			return nil, common.NewNonRetriableCtrlError("Rule '%s' appears twice in selections (enableRules or disableRules or manualRules)", selection.Name)
+		}
+		rule := &cmpv1alpha1.CustomRule{}
+		ruleKey := types.NamespacedName{Name: selection.Name, Namespace: tp.Namespace}
+		err := r.Client.Get(context.TODO(), ruleKey, rule)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return nil, common.NewNonRetriableCtrlError("Fetching rule: %w", err)
+			}
+			return nil, err
+		}
+
+		// Make sure all CustomRule has ScannerType as CEL as we only support CEL at this time
+		if rule.Spec.ScannerType != cmpv1alpha1.ScannerTypeCEL {
+			return nil, common.NewNonRetriableCtrlError("CustomRule '%s' has unsupported ScannerType: %s", rule.Name, rule.Spec.ScannerType)
+		}
+
+		// Check if the CustomRule is ready (validated)
+		if rule.Status.Phase != cmpv1alpha1.CustomRulePhaseReady {
+			if rule.Status.Phase == cmpv1alpha1.CustomRulePhaseError {
+				return nil, common.NewNonRetriableCtrlError("CustomRule '%s' has validation errors: %s", rule.Name, rule.Status.ErrorMessage)
+			}
+			// If the rule is not ready yet (perhaps still being validated), return a retriable error
+			// This will cause the TailoredProfile to be requeued
+			// Note: plain errors are retriable by default in this codebase
+			return nil, fmt.Errorf("CustomRule '%s' is not ready yet", rule.Name)
+		}
+
+		rules[selection.Name] = rule
+	}
+	return rules, nil
+}
+
+// validateNoMixedRuleTypes ensures that a TailoredProfile doesn't mix CustomRules and regular Rules
+func (r *ReconcileTailoredProfile) validateNoMixedRuleTypes(tp *cmpv1alpha1.TailoredProfile, customRules map[string]*cmpv1alpha1.CustomRule) error {
```

Comment:
We can probably consolidate this function or try to simplify it. We have two utility methods to get the CustomRule and Rule selections from a TailoredProfile. If both of those lists are not empty - then we have an error condition.

We can potentially do that without iterating the rule list again.